vid/src/avid_m/proofs.rs

```diff
@@ -0,0 +1,224 @@
+//! This module implements encoding proofs for the Avid-M Scheme.
+
+use std::collections::HashSet;
+
+use jf_merkle_tree::MerkleTreeScheme;
+use jf_utils::canonical;
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    avid_m::{
+        config::AvidMConfig, AvidMCommit, AvidMParam, AvidMScheme, AvidMShare, Config, MerkleProof,
+        MerkleTree, F,
+    },
+    VerificationResult, VidError, VidResult, VidScheme,
+};
+
+/// A proof of incorrect encoding.
+/// It consists of a witness vector, which is the claimed low degree polynomial,
+/// and a list of [`MalEncodingProofRawShare`]s. Each raw share claims that, in the Merkle tree committed by the provided commitment,
+/// some code word of the witness vector appears at the `index` position.
+/// If we have enough raw shares, and the Merkle commitment of the encoding of witness vector doesn't match the provided commitment,
+/// we assert that the provided commitment cannot open to be a correct encoding.
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct MalEncodingProof {
+    /// The witness vector
+    #[serde(with = "canonical")]
+    witness: Vec<F>,
+    /// Proof content.
+    raw_shares: Vec<MalEncodingProofRawShare>,
+}
+
+/// Proof content, see [`MalEncodingProof`].
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct MalEncodingProofRawShare {
+    /// The index of this raw share
+    pub index: usize,
+    /// A Merkle tree proof against the provided Merkle tree commitment.
+    #[serde(with = "canonical")]
+    pub mt_proof: MerkleProof,
+}
+
+impl AvidMScheme {
+    /// Generate a proof of incorrect encoding
+    /// See [`MalEncodingProof`] for details.
+    pub fn proof_of_incorrect_encoding(
+        param: &AvidMParam,
+        commit: &AvidMCommit,
+        shares: &[AvidMShare],
+    ) -> VidResult<MalEncodingProof> {
```

Comment:
I also wonder if we should further mention that caller of this code needs to check authenticity of `shares` and `commit`, i.e. whether they are signed by the disperser (or contains in the same VID proposal), since we don't check this internally.

Reply:
We did check that the shares should correspond to the commit.